contrib/verifybinaries/verify.py

```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Script for verifying Bitoin Core release binaries
+
+This script attempts to download the signature file SHA256SUMS.asc from
+bitcoincore.org and bitcoin.org and compares them.
+It first checks if the signature passes, and then downloads the files
+specified in the file, and checks if the hashes of these files match those
+that are specified in the signature file.
+The script returns 0 if everything passes the checks. It returns 1 if either
+the signature check or the hash check doesn't pass. If an error occurs the
+return value is >= 2.
+"""
+from hashlib import sha256
+import os
+import subprocess
+import sys
+from textwrap import indent
+
+WORKINGDIR = "/tmp/bitcoin_verify_binaries"
+HASHFILE = "hashes.tmp"
+HOST1 = "https://bitcoincore.org"
+HOST2 = "https://bitcoin.org"
+VERSIONPREFIX = "bitcoin-core-"
+SIGNATUREFILENAME = "SHA256SUMS.asc"
+
+
+def parse_version_string(version_str):
+    if version_str.startswith(VERSIONPREFIX):  # remove version prefix
+        version_str = version_str[len(VERSIONPREFIX):]
+
+    parts = version_str.split('-')
+    version_base = parts[0]
+    version_rc = ""
+    version_os = ""
+    if len(parts) == 2:  # "<version>-rcN" or "version-platform"
+        if "rc" in parts[1]:
+            version_rc = parts[1]
+        else:
+            version_os = parts[1]
+    elif len(parts) == 3:  # "<version>-rcN-platform"
+        version_rc = parts[1]
+        version_os = parts[2]
+
+    return version_base, version_rc, version_os
+
+
+def download_with_wget(remote_file, local_file=None):
+    if local_file:
+        wget_args = ['wget', '-O', local_file, remote_file]
+    else:
+        # use timestamping mechanism if local filename is not explicitely set
+        wget_args = ['wget', '-N', remote_file]
+
+    result = subprocess.run(wget_args,
+                            stderr=subprocess.STDOUT, stdout=subprocess.PIPE)
+    return result.returncode == 0, result.stdout.decode().rstrip()
+
+
+def files_are_equal(filename1, filename2):
+    with open(filename1, 'rb') as file1:
+        contents1 = file1.read()
+    with open(filename2, 'rb') as file2:
+        contents2 = file2.read()
+    return contents1 == contents2
+
+
+def verify_with_gpg(signature_filename, output_filename):
+    result = subprocess.run(['gpg', '--yes', '--decrypt', '--output',
```

Comment:
Obviously unrelated, but I wondered why we'd use `--decrypt` to `--verify` when the files aren't encrypted

Reply:
Oh, I see it is used to protect against attacks where some parts of the file are not covered by the signature.

Reply:
What attack are you referring to?

My understanding is that `--decrypt` is used here because it makes `gpg` output the signed message to `STDOUT` (or the file specified by `--output`), without the header and signature, while `--verify` won't give you that.

I commonly use it like that: `curl https://foo.bar/SHA256SUMS.asc | gpg --decrypt - | sha256sum -c -`.

Yes, it is kind of strange that its called `decrypt` when in reality it doesn't decrypt anything...

Reply:
Another thing to note is that `gpg` checks the signature against any of the public keys in the local keyring. This can be improved by creating a keyring with just Bitcoin Core's signing key and verifying against that:

```bash
$ gpg --export 90C8019E36C2E964 |  gpg --no-default-keyring --keyring ./bitcoincore.keyring --import
$ gpg --no-default-keyring --keyring ./bitcoincore.keyring --decrypt SHA256SUMS.asc
```

(Note the `./` for `--keyring` -- the keyring file will be created in `~/.gnupg` instead of the PWD without it)

I would also consider switching to [`gpgv`](https://www.gnupg.org/documentation/manuals/gnupg/gpgv.html), which is part of the gnupg suite but does only signature verification, is much smaller/simpler, can be installed independently and is more widely available (i.e. its usually there on these "slim" docker images).

Reply:
You can append/prepend any data to any historically valid signature and it will happily "verify". E.g. 


```
gpg --verify SHA256SUMS.asc && tail -2 SHA256SUMS.asc 
gpg: Signature made Wed 03 Jun 2020 11:59:52 AM CEST
gpg:                using RSA key 90C8019E36C2E964
gpg: Good signature ...
ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff  bitcoin-22.0-windows.zip
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa  bitcoin-22.0-linux.zip